src/System.Device.Gpio/System/Device/Gpio/Drivers/Libgpiod/LibGpiodDriverFactory.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Device.Gpio.System.Device.Gpio.Drivers.Libgpiod.V1;
+using System.Device.Gpio.System.Device.Gpio.Drivers.Libgpiod.V2;
+using System.IO;
+using System.Linq;
+
+namespace System.Device.Gpio.System.Device.Gpio.Drivers.Libgpiod;
+
+internal static class LibGpiodDriverFactory
+{
+    private const int _defaultDriverVersion = 1;
+
+    private const string _librarySearchPattern = "libgpiod.so*";
+    private const string _v0 = "libgpiod.so.0";
+    private const string _v1_0 = "libgpiod.so.1";
+    private const string _v1_1 = "libgpiod.so.2";
+    private const string _v2 = "libgpiod.so.3";
+
+    private static readonly string[] _librarySearchPaths = { "/lib", "/usr/lib", "/usr/local/lib" };
+
+    private static readonly object _lockObject = new();
+    private static int? _driverVersionToLoad;
+
+    /// <summary>
+    /// Tries with best effort to find installed libgpiod libraries and loads the latest version.
+    /// <see href="https://learn.microsoft.com/en-us/dotnet/standard/native-interop/native-library-loading#custom-import-resolver"/>
+    /// </summary>
+    /// <remarks> On NET Core 3.1 or NET5+ SetDllImportResolver would be a better solution</remarks>
+    public static GpioDriver Create(int chipNumber)
+    {
+        lock (_lockObject)
+        {
+            if (_driverVersionToLoad.HasValue)
+            {
+                return InstantiateDriver(_driverVersionToLoad.Value, chipNumber);
+            }
+
+            _driverVersionToLoad = GetDriverVersionToLoad();
+            return InstantiateDriver(_driverVersionToLoad.Value, chipNumber);
+        }
+    }
+
+    private static GpioDriver InstantiateDriver(int driverVersionToLoad, int chipNumber)
+    {
+        return driverVersionToLoad switch
+        {
+            1 => new LibGpiodV1Driver(chipNumber),
+            2 => new LibGpiodV2Driver(chipNumber),
+            _ => throw new ArgumentOutOfRangeException(nameof(driverVersionToLoad), $"Unexpected libgpiod driver version: '{driverVersionToLoad}'")
+        };
+    }
+
+    private static int GetDriverVersionToLoad()
+    {
+        HashSet<string> foundLibraryFiles = new();
+
+        foreach (string searchPath in _librarySearchPaths)
```

Comment:
I honestly don't love this approach of having some magic places where we look for things, it is code that will likely cause issues in the future. Is there a PInvoke present on all versions that return the version number of the library? If so, I'd rather us do a discovery attempt where we try PInvoking to get that info, and similar to the previous approach use NativeLibrary.TryLoad to do the discovery. Having a list like this is very fragile, and likely to be broken for some scenarios.

Reply:
I don't like it either but it was the best thing I could come up with. It basically is based on the file system hierarchy standard of linux, that most distros follow.
If I could choose the framework to do this, I would, but `NativeLibrary` does not seem to be available. Is this because netstandard 2.0 is targeted?

Reply:
maybe, honestly I don't think Netstandard2.0 is getting us much here, so we should probably consider dropping it. Main place where it is still useful is for devices like raspberry pi nano that can run Mono but not .NET Core.

Reply:
I see following options:
1. on .NET standard use the loading approach of this class, on  >= NET Core 3.0 use NativeLibrary
2. drop netstandard support and only use NativeLibrary (unsure about impact)

With either option we should consolidate the code of LibgpiodV1.Interop static ctor with the new approach.